Updates to terrain world component docs. (#1884)
o3de/o3de.org: content/docs/user-guide/components/reference/terrain/world-renderer.md

You must add this component to the *Level* entity, the parent of all entities in an **Open 3D Engine (O3DE)** level.
{{< /important >}}

## Usage 
The terrain renderer component allows terrain to be visualized in the world and controls various global properties that apply to terrain everywhere. 

#### Mesh Configuration
These properties control the mesh density and blending of terrain LODs (level of detail) and have a very direct impact on vertex performance. The density of the first terrain
LOD is set by the query resolution in the [Terrain World](/docs/user-guide/components/reference/terrain/world) component, and **First LOD distance** controls how far to render at this
density. Each subsequent LOD will render for twice the distance at half the resolution all the way until **Mesh render distance** is met. This means that **First LOD distance** actually
has a much greater impact on total triangle count than the total rendering distance.
#### Detail Configuration
These settings impact how far detail materials render, how they fade out, and how they blend with each other. Most of these settings have little impact on performance, however a larger

> render distance here will require more surface data points to be queried as you move around the world and also take more memory.

Reviewer: Doesn't querying more surface data points directly impact the overall performance?

Author: Which is why I said 

> however a larger render distance here will require more surface data points to be queried as you move around the world and also take more memory.

So the performance is impacted _a little_ as you move the camera, but it's pretty negligible when not moving, and even overall it's not a huge cost to query 1000 surface points every few frames. Doesn't really compare to vertex cost of dense meshes.

Reviewer: At a minimum, I'd suggest rephrasing this as "Most of these settings have little impact on per-frame rendering performance", since there is a noticeable performance impact when moving the camera right now.

#### Clipmap Configuration

> Turning on clipmaps causes all the material blending to be baked into textures for a substantial gain in performance, but also comes with a substantial increase in memory cost.

Reviewer: From what I've seen so far, I'm not sure it's true that we can say there's a substantial gain in performance. Steady-state rendering performance appears to be faster, but if you fly rapidly around the world, the clipmaps currently seem to have far worse performance than the forward rendering version.

Reviewer: Suggestion: This could also benefit from a brief description of what clipmaps are and how they work. I believe Jiawei posted a helpful diagram on a PR showing what's happening here.

Author: I was intentionally light on info with clip maps since I don't know all the details of its implementation. It'd be great for that to be fleshed out by @jiaweig-amzn :). As for the statement about performance, clipmaps _should_ be substantially faster in most normal applications. I'm sure there are many opportunities to make them more performant as time goes on, but maybe we need to be more precise about their expected performance in different applications.

Reviewer 2: I have sketched this out. Feel free to modify any word. The sentence below in brackets is probably too much detail. We can remove that.
Clipmaps are a stack of baked textures centered at the view and covering different distances at different resolutions. Detail materials and macro materials are backed into 2 sets of clipmaps which will take their place when the terrain is rendered. (In O3DE implementation, clipmaps have a uniform size, and the least index stands for the highest resolution clipmap.) Using clipmap will have a substantial gain in performance at the cost of texture memories. Clipmap configuration can be used to find a balance between the performance gain and the memory cost.

density. Each subsequent LOD will render for twice the distance at half the resolution all the way until **Mesh render distance** is met. This means that **First LOD distance** actually
has a much greater impact on total triangle count than the total rendering distance.

| Property | Description | Values | Default |
| - | - | - | - |
| Mesh render distance | The distance from the camera that terrain meshes will render in meters. | 1.0 to 100000.0 | 4096.0 |
| First LOD distance | The distance from the camera that the first LOD renders to. Subsequent LODs will be at double the distance from the previous LOD. | 1.0 to 10000.0 | 128.0 |
| Continuous LOD (CLOD) | Enables the use of continuous level of detail, which smoothly blends geometry between terrain LODs. | Boolean | true |
| CLOD Distance | Distance in meters over which the first LOD will blend into the next LOD. Subsequent LOD blend distances will double with each LOD for a consistent visual appearance. <br><br>  {{< note >}} Generally this should be less than about 25% of **First LOD distance**. If set too low, seams may appear in the terrain. {{< /note >}} | 0.0 to 1000.0 | 16.0 |

### Detail Configuration

> These settings impact how far detail materials render, how they fade out, and how they blend with each other. Most of these settings have little impact on performance, however a larger
> render distance here will require more surface data points to be queried as you move around the world and also take more memory.

Reviewer 3: ```suggestion
 These settings impact how far detail materials render, how they fade out, and how they blend with each other. Most of these settings have little impact on performance. However, a larger render distance requires more surface data points to be queried. As you move around the world, this uses more memory.    
```

Author: The way the last two sentences read changes the meaning a bit. More surface points are queried when you move around the world, and it takes more memory (regardless of if you're moving or not). Changing to 
"However, a larger render distance requires more surface data points to be queried as you move around the world, and it also uses more memory."

| Property | Description | Values | Default |
| - | - | - | - |

> | Height based texture blending | When turned on, detail materials will use the height texture to aid with blending. | Boolean | False |

Reviewer 3: ```suggestion
| **Height based texture blending** | If enabled, detail materials use the height texture for blending. | Boolean | False |
```

> | Detail material render distance (m) | The distance from the camera that the detail material will render in meters. | 1.0 to 2048.0 | 512.0 |

Reviewer 3: ```suggestion
| **Detail material render distance** | The distance in meters from the camera that the detail material renders to. | 1.0 to 2048.0 | 512.0 |
```

> | Detail material fade distance (m) | The distance over which the detail material will fade out into the macro material. | 0.0 to 2048.0 | 64.0 |

Reviewer 3: ```suggestion
| **Detail material fade distance** | The distance in meters over which the detail material fades out into the macro material. | 0.0 to 2048.0 | 64.0 |
```

> | Detail material scale | The scale at which all detail materials are rendered at. | 0.0001 to 10000.0 | 1.0 |

Reviewer 3: ```suggestion
| **Detail material scale** | The scale that all detail materials render at. | 0.0001 to 10000.0 | 1.0 |
```

### Clipmap Configuration

> Clipmaps are a stack of baked textures centered at the view and covering different distances at different resolutions. Detail materials and macro materials are baked into 2 sets of clipmaps which will take their place when the terrain is rendered. Using clipmaps will have a substantial gain in performance at the cost of texture memory. Clipmap configuration can be used to find a balance between the performance gain and the memory cost.

Reviewer 3: - Wdym by "centered at the view"?
- In my suggested rephrasing, is it accurate to say that the detail and macro materials "replace the clipmaps"?
```suggestion
_Clipmaps_ are a stack of baked textures that are centered at the view. They include baked textures at different distances and at different resolutions. Detail materials and macro materials are baked into two sets of clipmaps. They replace the clipmaps when rendering the terrain. Using clipmaps improves performance, but at the cost of texture memory. You can adjust the clipmap configuration to find a balance between the performance gain and the memory cost. 
```

Author: This needed a bit more work, I messaged you directly with a suggestion.

| Property | Description | Values | Default |
| - | - | - | - |

> | Clipmap Enabled | When turned on, renders terrain materials with clipmaps instead of rendering the materials directly every frame. | Boolean | false

Reviewer 3: ```suggestion
| **Clipmap Enabled** | If enabled, renders terrain materials with clipmaps, instead of rendering the materials directly every frame. | Boolean | False
```

> | Clipmap image size | The size of the clipmap image in each layer. | 512, 1024, or 2048 | 1024

Reviewer 3: ```suggestion
| **Clipmap image size** | The size of the clipmap image in each layer. | 512, 1024, or 2048 | 1024
```

> | Macro clipmap max resolution: texels/m | The resolution of the highest resolution clipmap in the stack. | 0.1 to 100.0 | 2.0

Reviewer 3: Is it possible to remove "texels/m" in the property name of the component card?
```suggestion
| **Macro clipmap max resolution: texels/m** | The resolution of the highest resolution clipmap in the stack. | 0.1 to 100.0 | 2.0
```

> | Detail clipmap max resolution: texels/m | The resolution of the highest resolution clipmap in the stack. | 10.0 to 4096.0 | 2048.0

Reviewer 3: ```suggestion
| **Detail clipmap max resolution: texels/m** | The resolution of the highest resolution clipmap in the stack. | 10.0 to 4096.0 | 2048.0
```

> | Macro clipmap scale base | The scale base between two adjacent clipmap layers. For example, 3 means the (n+1)th clipmap covers 3^2 = 9 times the area covered by the nth clipmap. | 1.1 to 10.0 | 2.0

Reviewer 3: ```suggestion
| **Macro clipmap scale base** | The scale base between two adjacent clipmap layers of macro materials. For example, 3 means that the (n+1)<sup>th</sup> clipmap covers 3<sup>2</sup> = 9 times the area that's covered by the n<sup>th</sup> clipmap. | 1.1 to 10.0 | 2.0
```

> | Detail clipmap scale base | The scale base between two adjacent clipmap layers. For example, 3 means the (n+1)th clipmap covers 3^2 = 9 times the area covered by the nth clipmap. | 1.1 to 10.0 | 2.0

Reviewer 3: ```suggestion
| **Detail clipmap scale base** | The scale base between two adjacent clipmap layers of detail materials. For example, 3 means that the (n+1)<sup>th</sup> clipmap covers 3<sup>2</sup> = 9 times the area that's covered by the n<sup>th</sup> clipmap. | 1.1 to 10.0 | 2.0
```

> | Macro clipmap margin size: texels | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates but also results in lower resolution clipmaps rendering closer to the camera. | 1 to 16 | 4

Reviewer 3: ```suggestion
| **Macro clipmap margin size: texels** | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates, but also results in lower resolution clipmaps when rendering closer to the camera. | 1 to 16 | 4
```

> | Detail clipmap margin size: texels | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates but also results in lower resolution clipmaps rendering closer to the camera. | 1 to 16 | 4

Reviewer 3: ```suggestion
| **Detail clipmap margin size: texels** | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates, but also results in lower resolution clipmaps when rendering closer to the camera. | 1 to 16 | 4
```

### Clipmap Configuration

| Property | Description | Values | Default |
| - | - | - | - |
| **Clipmap Enabled** | If enabled, renders terrain materials with clipmaps, instead of rendering the materials directly every frame. | Boolean | false
| **Clipmap image size** | The size of the clipmap image in each layer. | 512, 1024, or 2048 | 1024
| **Macro clipmap max resolution: texels/m** | The resolution of the highest resolution clipmap in the stack. | 0.1 to 100.0 | 2.0
| **Detail clipmap max resolution: texels/m** | The resolution of the highest resolution clipmap in the stack. | 10.0 to 4096.0 | 2048.0
| **Macro clipmap scale base** | The scale base between two adjacent clipmap layers of macro materials. For example, 3 means that the (n+1)<sup>th</sup> clipmap covers 3<sup>2</sup> = 9 times the area that's covered by the n<sup>th</sup> clipmap. | 1.1 to 10.0 | 2.0
| **Detail clipmap scale base** | The scale base between two adjacent clipmap layers of detail materials. For example, 3 means that the (n+1)<sup>th</sup> clipmap covers 3<sup>2</sup> = 9 times the area that's covered by the n<sup>th</sup> clipmap. | 1.1 to 10.0 | 2.0

> | **Macro clipmap margin size: texels** | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates, but also results in lower resolution clipmaps when rendering closer to the camera. | 1 to 16 | 4

Reviewer 4: ```suggestion
| **Macro clipmap margin size: texels** | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates, but also results in lower resolution clipmaps when rendering closer to the camera. | 1 - 16 | `4` |
```

> | **Detail clipmap margin size: texels** | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates, but also results in lower resolution clipmaps when rendering closer to the camera. | 1 to 16 | 4

Reviewer 4: ```suggestion
| **Detail clipmap margin size: texels** | The margin of the clipmap beyond the visible data. Increasing the margins results in less frequent clipmap updates, but also results in lower resolution clipmaps when rendering closer to the camera. | 1 - 16 | `4` |
```

## Properties

![Terrain World Renderer component interface.](/images/user-guide/components/reference/terrain/terrain-world-renderer-A.png)

| Property | Description | Values | Default |
| - | - | - | - |
| **Mesh Configuration** |

> | Mesh render distance | The distance from the camera that terrain meshes will render. | 1.0 to 100000.0 | 4096.0 |

Reviewer: Suggestion: "will render in meters."

> | First LOD distance | The distance from the camera that the first Lod renders to. Subsequent LODs will be at double the distance from the previous LOD. | 1.0 to 10000.0 | 128.0 |

Reviewer: Nit: "Lod" -> "LOD"

| Continuous LOD (CLOD) | Enables the use of continuous level of detail, which smoothly blends geometry between terrain lods. | Boolean | true |

> | CLOD Distance | Distance in meters over which the first lod will blend into the next lod. Subsequent lod blend distances will double with each lod for a consistent visual appearance. | 0.0 to 1000.0 | 16.0

Reviewer: Nit: "lod" -> "LOD"

| |
| **Detail Material Configuration** |

> | Height based texture blending | When turned on, detail materials will use the height texture to aid with blending. | Boolean | 0.0 |

Reviewer: Default is "False", not "0.0". (Side note - should we change this to default to True?)

Author: We probably should set this to true, but it's false right now. I'll update the wording.
